Testscripts/Linux/handle_offline_cpu.sh

```diff
@@ -0,0 +1,148 @@
+				echo 0 > $_device/channels/$_vmbus_ch/cpu
+				sleep 1
+				# validate the change of that cpu id
+				_cpu_id=$(cat $_device/channels/$_vmbus_ch/cpu)
+				if [ $_cpu_id = 0 ]; then
+					LogMsg "Successfully set the vmbus channel $_vmbus_ch's cpu to 0"
+				else
+					LogErr "Failed to set the vmbus channel $_vmbus_ch's cpu to 0. Expected 0, but found $_cpu_id"
+					FailedCount=$((FailedCount+1))
+				fi
+			fi
+			sleep 1
+		done < channel_vp_mapping
+	done
+	LogMsg "The list of target CPU: ${idle_cpus[@]}"
+
+	# #######################################################################################
+	# The previous step sets all channels' cpu to 0, so the rest of non-zero cpu can be offline
+	# Select a CPU number where does not associate to vmbus channels from idle_cpus array
+	LogMsg "Set all online cpus to offline."
+	for id in ${idle_cpus[@]}; do
+		state=$(cat /sys/devices/system/cpu/cpu$id/online)
+		if [ $state = 1 ]; then
+			# Set cpu to offline
+			echo 0 > /sys/devices/system/cpu/cpu$id/online
+			sleep 1
+			LogMsg "Set the cpu $id offline"
+			post_state=$(cat /sys/devices/system/cpu/cpu$id/online)
+			if [ $post_state = 0 ]; then
+				LogMsg "Successfully verified the cpu $id offline"
+			else
+				LogErr "Failed to verify the cpu $id state. Expected 0, found $post_state"
+				FailedCount=$((FailedCount+1))
+			fi
+		fi
+	done
+
+	# ########################################################################
+	# The previous steps set all cpus to offline, but all channels use cpu 0
+	# By using ethtool command, it sets new channels
+	# Verify if any offline cpu is assigned to new channels or not.
+	# Verify cpu should stay offline state.
+	LogMsg "Change vmbus channel numbers by ethtool command"
+	_ch_counts=$(ethtool -l eth0 | grep -i combined | tail -1 | cut -d ':' -f 2 | sed -e 's/^[[:space:]]*//')
+	LogMsg "Current channel counts: $_ch_counts"
+
+	# Set the random channel numbers to network device.
+	# max channel number is 64.
+	# Source: https://github.com/torvalds/linux/blob/master/drivers/net/hyperv/hyperv_net.h#L842
+	_new_counts=$(($RANDOM % 64))
+	((_new_counts=_new_counts+1))
+	if [[ $_new_counts -gt $max_cpu ]]; then
+		_new_counts=$max_cpu
+	fi
+
+	ethtool -L eth0 combined $_new_counts
+	sleep 1
+	LogMsg "Changed the channel numbers to $_new_counts"
+
+	if [ ! -d "$syn_net_adpt" ]; then
+		LogErr "Can not find the synthetic network adapter of vmbus sysfs path. The test failed."
+		FailedCount=$((FailedCount+1))
+	else
+		cp $syn_net_adpt/channel_vp_mapping new_channel_vp_mapping
+
+		while IFS=: read v c; do
+			LogMsg "Found vmbus channel: $v,		cpu: $c"
+			# Run test against all except cpu0.
+			if [ $c != 0 ]; then
+				# CPU is offline
+				_state=$(cat /sys/devices/system/cpu/cpu$c/online)
+				if [ $_state = 0 ]; then
+					LogMsg "Successfully verify the cpu $id offline"
+				else
+					LogErr "Failed to verify the cpu $id offline. Expected 0, found $post_state"
+					FailedCount=$((FailedCount+1))
```

Comment:
The other way around, right? (i.e., failure case is when the channel interrupt is assigned to an offline CPU)

Reply:
Fixed